File: sdks/typescript/src/proxy.ts
/**
 * Non-invasive JS Proxy that observes provider calls.
 *
 * The proxy never subclasses provider internals. Every property access is
 * forwarded; only an allowlist of terminal create method paths (per adapter)
 * is intercepted. Sub-namespaces on the way to a terminal stay wrapped so we
 * can reach the terminal method; everything else passes straight through, so
 * the proxy survives provider SDK version churn.
 *
 * The provider call ALWAYS executes first. Cost calc + enqueue happen after
 * the result resolves; shipping happens in the background. The synchronous
 * overhead on the hot path is a metadata object-merge plus an array push.
 */

import type { ProviderAdapter, StreamAccumulator } from "./adapters/base.js";
import type { CallSiteMeta, MetadataResolver } from "./metadata.js";
import type { PricingEngine } from "./pricing.js";
import type { TelemetryQueue } from "./telemetry.js";
import {
  type Event,
  type Usage,
  SDK_LANG,
  SDK_VERSION,
  makeUsage,
  newRequestId,
} from "./schema.js";

/** Shared dependencies for a wrapped client tree. */
export interface Recorder {
  adapter: ProviderAdapter;
  resolver: MetadataResolver;
  pricing: PricingEngine;
  queue: Pick<TelemetryQueue, "enqueue">;
  /**
   * Provider name resolved at wrap time; may differ from adapter.name for
   * OpenAI-compatible providers reached via baseURL. Defaults to adapter.name.
   */
  provider?: string;
}

function utcNowIso(): string {
  // RFC3339 UTC with millisecond precision (Date#toISOString already gives ms).
  return new Date().toISOString();
}

function pathEquals(a: readonly string[], b: readonly string[]): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

function isTerminal(path: readonly string[], adapter: ProviderAdapter): boolean {
  return adapter.terminalPaths.some((t) => pathEquals(path, t));
}

function isPrefixOfTerminal(
  path: readonly string[],
  adapter: ProviderAdapter,
): boolean {
  return adapter.terminalPaths.some(
    (t) => path.length < t.length && pathEquals(path, t.slice(0, path.length)),
  );
}

/** Wrap a target object in a Proxy that intercepts terminal create methods. */
function wrapNode(
  target: object,
  recorder: Recorder,
  path: readonly string[],
): unknown {
  return new Proxy(target, {
    get(obj, prop, receiver) {
      if (typeof prop !== "string") {
        return Reflect.get(obj, prop, receiver);
      }
      const value = Reflect.get(obj, prop, receiver);
      const newPath = [...path, prop];

      // Terminal create call: intercept.
      if (isTerminal(newPath, recorder.adapter) && typeof value === "function") {
        return makeWrappedCreate(
          value as (...a: unknown[]) => unknown,
          obj,
          recorder,
          newPath,
        );
      }

      // On the path to a terminal and the node is an object: keep wrapping so
      // we can reach the terminal method.
      if (
        isPrefixOfTerminal(newPath, recorder.adapter) &&
        value !== null &&
        (typeof value === "object" || typeof value === "function")
      ) {
        return wrapNode(value as object, recorder, newPath);
      }

      // Everything else passes straight through. Bind functions to the real
      // target so `this` is correct.
      if (typeof value === "function") {
        return value.bind(obj);
      }
      return value;
    },

    set(obj, prop, value, receiver) {
      return Reflect.set(obj, prop, value, receiver);
    },
  });
}

/** Split the call-site `costobs` field out of the provider options object. */
function splitOptions(
  args: unknown[],
): { callMeta: CallSiteMeta; forwardArgs: unknown[] } {
  if (args.length === 0) return { callMeta: {}, forwardArgs: args };
  const first = args[0];
  if (first === null || typeof first !== "object" || Array.isArray(first)) {
    return { callMeta: {}, forwardArgs: args };
  }
  const opts = first as Record<string, unknown>;
  if (!("costobs" in opts)) {
    return { callMeta: {}, forwardArgs: args };
  }
  // Strip `costobs` before forwarding so the provider never sees it.
  const { costobs, ...rest } = opts;
  const callMeta = (costobs ?? {}) as CallSiteMeta;
  const forwardArgs = [rest, ...args.slice(1)];
  return { callMeta, forwardArgs };
}

function optionsObject(args: unknown[]): Record<string, unknown> {
  const first = args[0];
  if (first !== null && typeof first === "object" && !Array.isArray(first)) {
    return first as Record<string, unknown>;
  }
  return {};
}

interface CallState {
  requestId: string;
  ts: string;
  operation: string;
  model: string;
  streaming: boolean;
  callMeta: CallSiteMeta;
  start: number;
}

function makeWrappedCreate(
  createFn: (...a: unknown[]) => unknown,
  thisArg: object,
  recorder: Recorder,
  path: readonly string[],
): (...a: unknown[]) => unknown {
  const adapter = recorder.adapter;
  const operation = adapter.operationFor(path);
  // Methods whose name ends in "Stream" (e.g. Gemini generateContentStream)
  // always return a stream regardless of any option flag.
  const isStreamingMethod = path[path.length - 1]!.endsWith("Stream");

  return function wrappedCreate(...args: unknown[]): unknown {
    const { callMeta, forwardArgs } = splitOptions(args);
    let options = optionsObject(forwardArgs);
    const streaming = isStreamingMethod || adapter.isStream(options);

    if (streaming && !isStreamingMethod) {
      const prepared = adapter.prepareStreamOptions(options);
      if (prepared !== options) {
        options = prepared;
        forwardArgs[0] = prepared;
      }
    }

    const state: CallState = {
      requestId: newRequestId(),
      ts: utcNowIso(),
      operation,
      model: adapter.modelFromOptions(options),
      streaming,
      callMeta,
      start: performanceNow(),
    };

    // --- the provider call: nothing of ours runs before this is invoked ---
    let result: unknown;
    try {
      result = createFn.apply(thisArg, forwardArgs);
    } catch (e) {
      emit(recorder, state, makeUsage(), {
        status: "error",
        errorType: errorName(e),
      });
      throw e;
    }

    // Promise-returning create (async clients, or sync clients that return a
    // Promise). Attach handlers without changing the returned value's identity
    // semantics beyond what the provider already does.
    if (isPromiseLike(result)) {
      return (result as Promise<unknown>).then(
        (resolved) => finishResult(resolved, recorder, state, adapter),
        (err) => {
          emit(recorder, state, makeUsage(), {
            status: "error",
            errorType: errorName(err),
          });
          throw err;
        },
      );
    }

    return finishResult(result, recorder, state, adapter);
  };
}

function finishResult(
  result: unknown,
  recorder: Recorder,
  state: CallState,
  adapter: ProviderAdapter,
): unknown {
  if (state.streaming && isAsyncIterable(result)) {
    return wrapAsyncIterable(result, recorder, state, adapter);
  }
  if (state.streaming && isSyncIterable(result)) {
    return wrapSyncIterable(result, recorder, state, adapter);
  }
  const usage = adapter.parseUsage(result, state.operation);
  emit(recorder, state, usage, {});
  return result;
}

function performanceNow(): number {
  return typeof performance !== "undefined" ? performance.now() : Date.now();
}

function errorName(e: unknown): string {
  if (e instanceof Error) return e.name || e.constructor.name;
  if (e !== null && typeof e === "object") return e.constructor.name;
  return "Error";
}

function isPromiseLike(v: unknown): v is PromiseLike<unknown> {
  return (
    v !== null &&
    (typeof v === "object" || typeof v === "function") &&
    typeof (v as { then?: unknown }).then === "function"
  );
}

function isAsyncIterable(v: unknown): v is AsyncIterable<unknown> {
  return (
    v !== null &&
    typeof v === "object" &&
    typeof (v as { [Symbol.asyncIterator]?: unknown })[Symbol.asyncIterator] ===
      "function"
  );
}

function isSyncIterable(v: unknown): v is Iterable<unknown> {
  return (
    v !== null &&
    typeof v === "object" &&
    typeof (v as { [Symbol.iterator]?: unknown })[Symbol.iterator] === "function"
  );
}

interface EmitOpts {
  status?: "ok" | "error";
  errorType?: string;
}

function emit(
  recorder: Recorder,
  state: CallState,
  usage: Usage,
  opts: EmitOpts,
): void {
  const latencyMs = Math.round(performanceNow() - state.start);
  const meta = recorder.resolver.merge(state.callMeta);
  const adapter = recorder.adapter;
  const provider =
    adapter.providerForCall?.(state.model, {}) ??
    recorder.provider ??
    adapter.name;
  // Normalized for both pricing and the stamped event (strips routing
  // prefixes like "anthropic/claude-...").
  const model = adapter.normalizeModel?.(state.model) ?? state.model;
  const cost = recorder.pricing.cost(usage, provider, model, {
    operation: state.operation,
  });

  const event: Event = {
    request_id: state.requestId,
    ts: state.ts,
    provider,
    model,
    operation: state.operation,
    stream: state.streaming,
    status: opts.status ?? "ok",
    error_type: opts.errorType ?? "",
    environment: meta.environment ?? "",
    team: meta.team ?? "",
    service: meta.service ?? "",
    customer_id: meta.customer_id ?? "",
    user_id: meta.user_id ?? "",
    trace_id: meta.trace_id ?? "",
    feature: meta.feature ?? "",
    prompt_key: meta.prompt_key ?? "",
    prompt_version: meta.prompt_version ?? "",
    tags: meta.tags,
    input_tokens: usage.input_tokens,
    cached_input_tokens: usage.cached_input_tokens,
    output_tokens: usage.output_tokens,
    reasoning_tokens: usage.reasoning_tokens,
    tool_tokens: usage.tool_tokens,
    total_tokens: usage.total_tokens,
    audio_seconds: usage.audio_seconds,
    characters: usage.characters,
    image_count: usage.image_count,
    image_tiles: usage.image_tiles,
    cost_usd: cost.toNumber(),
    pricing_version: recorder.pricing.pricingVersion,
    latency_ms: latencyMs,
    sdk_lang: SDK_LANG,
    sdk_version: SDK_VERSION,
  };
  recorder.queue.enqueue(event);
}

/**
 * Wrap an async-iterable stream: tee usage as chunks flow, emit exactly once
 * when iteration completes OR is broken early (try/finally around the
 * passthrough). The returned object preserves the original's async-iterable
 * interface and forwards other properties.
 */
function wrapAsyncIterable(
  stream: AsyncIterable<unknown>,
  recorder: Recorder,
  state: CallState,
  adapter: ProviderAdapter,
): AsyncIterable<unknown> {
  const acc: StreamAccumulator = adapter.newStreamAccumulator(state.model);
  let emitted = false;
  const finish = (opts: EmitOpts = {}): void => {
    if (emitted) return;
    emitted = true;
    emit(recorder, state, acc.result(), opts);
  };

  async function* generator(): AsyncGenerator<unknown> {
    try {
      for await (const chunk of stream) {
        try {
          acc.feed(chunk);
        } catch {
          // never let usage parsing break the user's stream
        }
        yield chunk;
      }
      finish();
    } catch (e) {
      finish({ status: "error", errorType: errorName(e) });
      throw e;
    } finally {
      // Covers early `break` (generator.return) — emit once if not already.
      finish();
    }
  }

  return makeStreamProxy(
    stream,
    generator(),
    Symbol.asyncIterator,
  ) as AsyncIterable<unknown>;
}

/** Sync-iterable variant (some clients return a sync iterator for streams). */
function wrapSyncIterable(
  stream: Iterable<unknown>,
  recorder: Recorder,
  state: CallState,
  adapter: ProviderAdapter,
): Iterable<unknown> {
  const acc: StreamAccumulator = adapter.newStreamAccumulator(state.model);
  let emitted = false;
  const finish = (opts: EmitOpts = {}): void => {
    if (emitted) return;
    emitted = true;
    emit(recorder, state, acc.result(), opts);
  };

  function* generator(): Generator<unknown> {
    try {
      for (const chunk of stream) {
        try {
          acc.feed(chunk);
        } catch {
          // ignore
        }
        yield chunk;
      }
      finish();
    } catch (e) {
      finish({ status: "error", errorType: errorName(e) });
      throw e;
    } finally {
      finish();
    }
  }

  return makeStreamProxy(stream, generator(), Symbol.iterator) as Iterable<unknown>;
}

/**
 * Proxy that exposes the teeing (async-)generator for iteration but forwards
 * every other property to the original stream object (so `.controller`,
 * `.response`, `.tee()`, etc. still work).
 */
function makeStreamProxy(
  original: object,
  gen: Iterator<unknown> | AsyncIterator<unknown>,
  iterSymbol: symbol,
): object {
  return new Proxy(original, {
    get(obj, prop, receiver) {
      if (prop === iterSymbol) {
        return () => gen;
      }
      const value = Reflect.get(obj, prop, receiver);
      if (typeof value === "function") return value.bind(obj);
      return value;
    },
  });
}

/** Create the top-level observed client proxy. */
export function createObservedClient(client: object, recorder: Recorder): unknown {
  return wrapNode(client, recorder, []);
}

File: sdks/python/costobs/proxy.py
"""Non-invasive proxy that observes provider calls.

The proxy never subclasses provider internals. It forwards every attribute
access through ``__getattr__`` and only intercepts an allowlist of terminal
method paths (per adapter). Everything else passes straight through, so the
proxy survives provider SDK version churn.

The provider call ALWAYS executes first. Cost calc + enqueue happen after the
response returns; shipping happens on a background thread. The synchronous
overhead added on the hot path is a metadata dict-merge plus a queue
``put_nowait``.
"""

from __future__ import annotations

import inspect
import time
from datetime import datetime, timezone
from typing import Any, Tuple

from . import __version__
from .adapters import ProviderAdapter
from .metadata import MetadataResolver
from .pricing import PricingEngine
from .schema import Event, Usage, new_request_id
from .telemetry import TelemetryQueue

# Call-site kwargs the SDK consumes and strips before forwarding.
_COSTOBS_KWARGS = ("feature", "prompt_key", "prompt_version", "costobs_tags")


def _utc_now_iso() -> str:
    # RFC3339 UTC with millisecond precision.
    dt = datetime.now(timezone.utc)
    return dt.strftime("%Y-%m-%dT%H:%M:%S.") + f"{dt.microsecond // 1000:03d}Z"


class _Recorder:
    """Captures shared dependencies for a wrapped client tree."""

    __slots__ = ("adapter", "resolver", "pricing", "queue", "provider")

    def __init__(
        self,
        adapter: ProviderAdapter,
        resolver: MetadataResolver,
        pricing: PricingEngine,
        queue: TelemetryQueue,
        provider: str = "",
    ):
        self.adapter = adapter
        self.resolver = resolver
        self.pricing = pricing
        self.queue = queue
        # Resolved at wrap time (may differ from adapter.name for
        # OpenAI-compatible providers reached via base_url).
        self.provider = provider or adapter.name


class ObservedClient:
    """Attribute-path proxy around a provider client (or sub-object)."""

    # Real attributes live in __dict__; everything else proxies.
    def __init__(self, target: Any, recorder: _Recorder, path: Tuple[str, ...] = ()):
        object.__setattr__(self, "_target", target)
        object.__setattr__(self, "_recorder", recorder)
        object.__setattr__(self, "_path", path)

    def __getattr__(self, name: str) -> Any:
        # __getattr__ only runs for names not found normally, so no recursion.
        target = object.__getattribute__(self, "_target")
        recorder = object.__getattribute__(self, "_recorder")
        path = object.__getattribute__(self, "_path")

        attr = getattr(target, name)
        new_path = path + (name,)

        # Is this exact path a terminal create call we should intercept?
        if new_path in recorder.adapter.terminal_paths and callable(attr):
            return _make_wrapped_create(attr, recorder, new_path)

        # Is this path a strict prefix of any terminal path? If so, keep
        # proxying so we can reach the terminal method. If it's a leaf that
        # isn't terminal, return it untouched.
        if callable(attr) or not _is_path_prefix(new_path, recorder.adapter.terminal_paths):
            # Methods / leaves that aren't terminal pass straight through.
            # But sub-namespaces (non-callable objects on the way to a
            # terminal) must stay wrapped.
            if _is_path_prefix(new_path, recorder.adapter.terminal_paths) and not callable(attr):
                return ObservedClient(attr, recorder, new_path)
            return attr

        # Non-callable object on the path to a terminal -> keep wrapping.
        return ObservedClient(attr, recorder, new_path)

    def __setattr__(self, name: str, value: Any) -> None:
        setattr(object.__getattribute__(self, "_target"), name, value)

    def __repr__(self) -> str:
        return f"ObservedClient({object.__getattribute__(self, '_target')!r})"


def _is_path_prefix(path: Tuple[str, ...], terminals: Tuple[Tuple[str, ...], ...]) -> bool:
    return any(len(path) < len(t) and t[: len(path)] == path for t in terminals)


def _split_kwargs(kwargs: dict) -> Tuple[dict, dict]:
    """Separate CostObs call-site kwargs from provider kwargs."""
    meta: dict = {}
    for key in _COSTOBS_KWARGS:
        if key in kwargs:
            meta[key] = kwargs.pop(key)
    return meta, kwargs


def _make_wrapped_create(create_fn, recorder: _Recorder, path: Tuple[str, ...]):
    adapter = recorder.adapter
    operation = adapter.operation_for(path)

    is_async = inspect.iscoroutinefunction(create_fn)

    if is_async:

        async def async_wrapped(*args, **kwargs):
            call_meta, provider_kwargs = _split_kwargs(dict(kwargs))
            streaming = adapter.is_stream(provider_kwargs)
            if streaming:
                provider_kwargs = adapter.prepare_stream_kwargs(provider_kwargs)
            model = adapter.model_from_kwargs(provider_kwargs)
            request_id = new_request_id()
            ts = _utc_now_iso()
            start = time.perf_counter()

            try:
                result = await create_fn(*args, **provider_kwargs)
            except Exception as e:
                _emit(
                    recorder, request_id, ts, operation, model, streaming,
                    Usage(), call_meta, start, status="error",
                    error_type=type(e).__name__,
                )
                raise

            if streaming:
                return _AsyncStreamProxy(
                    result, recorder, request_id, ts, operation, model, call_meta, start
                )

            usage = adapter.parse_usage(result, operation)
            _emit(recorder, request_id, ts, operation, model, streaming, usage, call_meta, start)
            return result

        return async_wrapped

    def sync_wrapped(*args, **kwargs):
        call_meta, provider_kwargs = _split_kwargs(dict(kwargs))
        streaming = adapter.is_stream(provider_kwargs)
        if streaming:
            provider_kwargs = adapter.prepare_stream_kwargs(provider_kwargs)
        model = adapter.model_from_kwargs(provider_kwargs)
        request_id = new_request_id()
        ts = _utc_now_iso()
        start = time.perf_counter()

        # --- the provider call: nothing of ours runs before this returns ---
        try:
            result = create_fn(*args, **provider_kwargs)
        except Exception as e:
            _emit(
                recorder, request_id, ts, operation, model, streaming,
                Usage(), call_meta, start, status="error",
                error_type=type(e).__name__,
            )
            raise

        if streaming:
            return _SyncStreamProxy(
                result, recorder, request_id, ts, operation, model, call_meta, start
            )

        usage = adapter.parse_usage(result, operation)
        _emit(recorder, request_id, ts, operation, model, streaming, usage, call_meta, start)
        return result

    return sync_wrapped


def _emit(
    recorder: _Recorder,
    request_id: str,
    ts: str,
    operation: str,
    model: str,
    stream: bool,
    usage: Usage,
    call_meta: dict,
    start: float,
    *,
    status: str = "ok",
    error_type: str = "",
) -> None:
    """Build the event and enqueue it. Off the hot path of the provider call."""
    latency_ms = int((time.perf_counter() - start) * 1000)
    meta = recorder.resolver.merge(call_meta)
    adapter = recorder.adapter
    provider = adapter.provider_for_call(model, call_meta) or recorder.provider
    # Normalized for both pricing and the stamped event (strips routing
    # prefixes like LiteLLM's "anthropic/claude-...").
    model = adapter.normalize_model(model)

    cost = recorder.pricing.cost(usage, provider, model, operation)

    event = Event(
        request_id=request_id,
        ts=ts,
        provider=provider,
        model=model,
        operation=operation,
        stream=stream,
        status=status,
        error_type=error_type,
        environment=meta.get("environment", ""),
        team=meta.get("team", ""),
        service=meta.get("service", ""),
        customer_id=meta.get("customer_id", ""),
        user_id=meta.get("user_id", ""),
        trace_id=meta.get("trace_id", ""),
        feature=meta.get("feature", ""),
        prompt_key=meta.get("prompt_key", ""),
        prompt_version=meta.get("prompt_version", ""),
        tags=meta.get("tags", {}),
        input_tokens=usage.input_tokens,
        cached_input_tokens=usage.cached_input_tokens,
        output_tokens=usage.output_tokens,
        reasoning_tokens=usage.reasoning_tokens,
        tool_tokens=usage.tool_tokens,
        total_tokens=usage.total_tokens,
        audio_seconds=usage.audio_seconds,
        image_count=usage.image_count,
        image_tiles=usage.image_tiles,
        cost_usd=float(cost),
        pricing_version=recorder.pricing.pricing_version,
        latency_ms=latency_ms,
        sdk_lang="python",
        sdk_version=__version__,
    )
    recorder.queue.enqueue(event.to_dict())


class _SyncStreamProxy:
    """Passthrough iterator that tees usage and emits exactly once at the end."""

    def __init__(self, stream, recorder, request_id, ts, operation, model, call_meta, start):
        self._stream = stream
        self._recorder = recorder
        self._request_id = request_id
        self._ts = ts
        self._operation = operation
        self._model = model
        self._call_meta = call_meta
        self._start = start
        self._acc = recorder.adapter.new_stream_accumulator(model)
        self._emitted = False

    def __iter__(self):
        try:
            for chunk in self._stream:
                try:
                    self._acc.feed(chunk)
                except Exception:
                    pass
                yield chunk
        finally:
            # Fires on full exhaustion AND on early break (GeneratorExit).
            self._finish()

    def _finish(self, status: str = "ok", error_type: str = ""):
        if self._emitted:
            return
        self._emitted = True
        _emit(
            self._recorder, self._request_id, self._ts, self._operation,
            self._model, True, self._acc.result(), self._call_meta, self._start,
            status=status, error_type=error_type,
        )

    # Forward context-manager protocol (OpenAI streams support `with`).
    def __enter__(self):
        if hasattr(self._stream, "__enter__"):
            self._stream.__enter__()
        return self

    def __exit__(self, exc_type, exc, tb):
        self._finish(
            status="error" if exc_type else "ok",
            error_type=exc_type.__name__ if exc_type else "",
        )
        if hasattr(self._stream, "__exit__"):
            return self._stream.__exit__(exc_type, exc, tb)
        return False

    def __getattr__(self, name):
        return getattr(self._stream, name)


class _AsyncStreamProxy:
    """Async passthrough iterator that tees usage and emits exactly once."""

    def __init__(self, stream, recorder, request_id, ts, operation, model, call_meta, start):
        self._stream = stream
        self._recorder = recorder
        self._request_id = request_id
        self._ts = ts
        self._operation = operation
        self._model = model
        self._call_meta = call_meta
        self._start = start
        self._acc = recorder.adapter.new_stream_accumulator(model)
        self._emitted = False

    async def __aiter__(self):
        try:
            async for chunk in self._stream:
                try:
                    self._acc.feed(chunk)
                except Exception:
                    pass
                yield chunk
        finally:
            self._finish()

    def _finish(self, status: str = "ok", error_type: str = ""):
        if self._emitted:
            return
        self._emitted = True
        _emit(
            self._recorder, self._request_id, self._ts, self._operation,
            self._model, True, self._acc.result(), self._call_meta, self._start,
            status=status, error_type=error_type,
        )

    async def __aenter__(self):
        if hasattr(self._stream, "__aenter__"):
            await self._stream.__aenter__()
        return self

    async def __aexit__(self, exc_type, exc, tb):
        self._finish(
            status="error" if exc_type else "ok",
            error_type=exc_type.__name__ if exc_type else "",
        )
        if hasattr(self._stream, "__aexit__"):
            return await self._stream.__aexit__(exc_type, exc, tb)
        return False

    def __getattr__(self, name):
        return getattr(self._stream, name)
